### PR TITLE
Crop rounding argument to control the conversion of anchors to integral values

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

## Category:

**New feature**

## Description:

- Adds a `rounding` argument to Crop operator, to control how the anchors are converted to an integral number
- Two modes, truncate and round (default). 

### Affected modules and functionalities:

- Crop family

### Key points relevant for the review:

- Documentation

### Tests:

- test_crop.py: test_crop_rounding
- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [x] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [x] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: CROP.10

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by jantonguirao** [comment]
!build

File: dali/operators/image/crop/crop_attr.cc [hunk 1]
```diff
@@ -67,7 +75,16 @@ window dimensions (argument ``crop``).)code",
...
-        0.0f, true);
+        0.0f, true)
+    .AddOptionalArg(
+        "rounding", R"code(Determines the rounding function used to convert the starting coordinate
+of the window to an integral value (see ``crop_pos_x``, ``crop_pos_y``, ``crop_pos_z``).
+
+Possible values are:
+
+* | ``"round"`` - Rounds to the nearest integer value, with halfway cases rounded away from zero.
+* | ``"truncate"`` - Discards the fractional part of the number (truncates towards zero).)code",
```
Comment (mzient) [inline dali/operators/image/crop/crop_attr.cc:86]:
```suggestion
* | ``"truncate"`` - Discards the fractional part of the number (truncates).)code",
```
or
```suggestion
* | ``"truncate"`` - Discards the fractional part of the number (rounds towards zero).)code",
```

**Review by banasraf — APPROVED** [review]

File: dali/operators/image/crop/crop_attr.cc [hunk 2]
```diff
@@ -108,6 +125,20 @@ CropAttr::CropAttr(const OpSpec& spec) {
   crop_y_norm_.resize(max_batch_size, 0.0f);
   crop_z_norm_.resize(max_batch_size, 0.0f);
   crop_window_generators_.resize(max_batch_size, {});
+
+  auto rounding = spec.GetArgument<std::string>("rounding");
+  if (rounding == "round") {
+    round_fn_ = [](float x) {
+      return std::roundf(x);
+    };
```
Comment (mzient) [inline dali/operators/image/crop/crop_attr.cc:133]:
```suggestion
    round_fn_ = std::roundf;
```
wouldn't this work?

File: dali/operators/image/crop/crop_attr.h [hunk 3]
```diff
@@ -58,6 +58,7 @@ class CropAttr {
   std::vector<float> crop_z_norm_;
   std::vector<CropWindowGenerator> crop_window_generators_;
   bool is_whole_image_ = false;
+  std::function<float(float)> round_fn_;
```
Comment (mzient) [inline dali/operators/image/crop/crop_attr.h:61]:
```suggestion
  std::function<int(float)> round_fn_;
```
Don't we want to return integers?
Also, are we good with just `float` (and not double)?

File: dali/operators/image/crop/crop_attr.cc [hunk 4]
```diff
@@ -200,7 +231,8 @@ TensorShape<> CropAttr::CalculateAnchor(const span<float>& anchor_norm,
     DALI_ENFORCE(anchor_norm[dim] >= 0.0f && anchor_norm[dim] <= 1.0f,
                  "Anchor for dimension " + std::to_string(dim) + " (" +
                      std::to_string(anchor_norm[dim]) + ") is out of range [0.0, 1.0]");
-    anchor[dim] = std::roundf(anchor_norm[dim] * (input_shape[dim] - crop_shape[dim]));
+    float anchor_f = anchor_norm[dim] * (input_shape[dim] - crop_shape[dim]);
+    anchor[dim] = round_fn_(anchor_f);
```
Comment (mzient) [inline dali/operators/image/crop/crop_attr.cc:235]: While we're at it, I'd bump precision to double. It's not very likely to happen, but we might run into some bizarre 1D case that would be more than 8M pixels wide.

**Review by mzient — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

**Pull request merged**